Persistent Bans Overhaul

ezekieltem opened the pull request:
I've started an attempt to overhaul Persistent Bans. I plan on adding a pbanlogs command, but wanted to get this pull request made so changes can be requested and potential oversights can be spotted.

I've kept support in for the old Persistent Bans system primarily for backwards-compatibility reasons, and should developers prefer to continue using the old system.

From what limited testing I can do (with my main and just an alt), I believe it functions correctly. Though I could have, and maybe likely, missed some cases.

Fun fact: Did you know the owner of an experience can be banned from their own experience? Both directly and indirect via Roblox's alt detection? Found that out while testing this.

ezekieltem committed:
Main Commit 1 of my attempt at a Persistent Bans Overhaul.

ezekieltem committed:
Merge branch 'master' of https://github.com/ezekieltem/Nexus-Admin

ezekieltem committed:
Rename pbans.legacy.luau to pbans-legacy.luau

ezekieltem commented:
Just realized I left a lot of my debug prints in, I'll have them removed by final commit. Which will likely be tomorrow afternoon (CST).

ezekieltem committed:
Final commit for Persistent Bans Overhaul (should be), before requests/revisions/oversights.

TheNexusAvenger commented:
This is really neat to see. It will be a few days before I can review it due to RDC, but I should have a review by the 10th.

TheNexusAvenger commented:
I've looked over this a bit and I like where this is going. However, there is a design decision I wanted to ask about: do you see any advantages to keeping legacy bans around as opposed to only having the migration to the new bans? In theory, no experience should be relying on the existing behavior of the DataStores.

Additional question: there are some lesser styling things that I would like to have changed before a release. Would you prefer I post them in a review and have you fix them, or merge the changes and then I'll handle them afterwards?

ezekieltem commented:
I kept the legacy bans in the event an experience does have separate scripts that do rely on it for whatever reason. The only possible advantage I could see, would be for developers who *do* rely on it, to not have to use a custom NexusAdmin module and lose access to automatic updates, assuming they are accessing PersistentBans via `Api.CommandData.PersistentBans`, and not calling the module directly. Overall, I view the choice as yours if it stays in or not.

As for styling things, go ahead and throw them at me and I'll try and integrate them.

TheNexusAvenger commented:
(Sorry for the delay; was busy with travel) I think it will be better long-term to drop the legacy bans and only maintain the new ban API. While the bans stuff was exposed, it was never documented and can't be assumed to exist long-term. That and the internal format was so limited (no auditability, no time information, no temp bans) that it shouldn't have been used for scripting.

Bit of a warning - I'll be busy up until the weekend and might not respond. After that, I'll post 1 review for any logic stuff (if any) and 1 review for the nitpick formatting stuff. After that, I will be holding the release of it for V.2.10.0 while I [hopefully] wait for access to the Server Authority early access to ensure commands like `fly` work.

ezekieltem committed:
Remove Legacy Persistent Bans, add check if enviornment is RobloxStudio

ezekieltem committed:
Minor Api doc updates (also actually add the Studio Environment Check)

ezekieltem committed:
Fix var name & remove intentional error

Realized I left an intentional error in that I had put for testing reasons.

I swear, every time I look at this I spot something.

TheNexusAvenger commented on src/Common/Configuration.luau:
```diff
@@ -76,6 +79,10 @@ function Configuration.new(ConfigurationTable: {[string]: any}): Configuration
     --Confirm all DEFAULT_COMMAND_CONFIGURATIONS exist within CommandConfigurations.
     for Name, Value in DEFAULT_COMMAND_CONFIGURATIONS do
         if self.CommandConfigurations[Name] ~= nil then continue end
+        if type(Value) == "string" and (Value :: string):sub(1,2) == "${" then
+            local Key = Value:sub(3,Value:len()-1)
+            Value = self[Key]
+        end
```
The idea of string replacing like this is clever, but I'm nervous about the cost to maintain for 1 feature now, and potentially more in the future if it gets expanded. My preference would be to add something like the following below:
```luau
if not self.CommandConfigurations.PersistentBanImmuneLevel then --Note: added Level to the end to be consistent with DefaultServerLockAdminLevel.
    self.CommandConfigurations.PersistentBanImmuneLevel = self.PersistentCommandsLevel
end
```

ezekieltem replied:
One thing that crosses my mind, is in the event that more command configurations are added that default to the value of a different configuration. Then depending on the amount, there could end up just being a mass about of lines that follow that just set specific command configurations.

Assuming I interpret what you are saying correctly, which I am seeing as something like this
```lua
    --Confirm all DEFAULT_COMMAND_CONFIGURATIONS exist within CommandConfigurations.
    for Name, Value in DEFAULT_COMMAND_CONFIGURATIONS do
        if self.CommandConfigurations[Name] ~= nil then continue end
        self.CommandConfigurations[Name] = Value
    end
    
    --DEFAULT_COMMAND_CONFIGURATIONS that default to other specified configurations.
    if not self.CommandConfigurations.PersistentBanImmuneLevel then
        self.CommandConfigurations.PersistentBanImmuneLevel = self.PersistentCommandsLevel
    end
```

TheNexusAvenger commented on src/IncludedCommands/Persistent/pban.luau:
```diff
@@ -1,46 +1,191 @@
+        local PersistentBans = require(script.Parent.Parent:WaitForChild("Resources"):WaitForChild("PersistentBans"))
         Api.CommandData.PersistentBans = PersistentBans.GetInstance(Api)
+
+        local function UserIdToPlayer(Id: number) : Player?
+            local Match
+            for _, Player in Players:GetPlayers() do
+                if Player.UserId == Id then
+                    Match = Player
+                    break
+                end
+            end
+            return Match
+        end
```
Can references to this be replaced with `Players:GetPlayerByUserId(Id)`?

ezekieltem replied:
I genuinely forgot that function existed.

TheNexusAvenger commented on src/IncludedCommands/Persistent/pban.luau:
```diff
@@ -1,46 +1,191 @@
+            local Filtered = {}
+            local PersistentBanImmune = Api.Configuration.CommandConfigurations.PersistentBanImmune
+            for i, Id in Ids do
+                local Player = UserIdToPlayer(Id)
+                local AdminLevel = -1
+                if Player then
+                    if Player == Executor then
+                        continue
+                    end
+                    AdminLevel = math.max(AdminLevel,Api.Authorization:YieldForAdminLevel(Player))
+                else
+                    if Api.Configuration.Admins[Id] then
+                        AdminLevel = math.max(AdminLevel,Api.Configuration.Admins[Id])
+                    end
+                    xpcall(function()
+                        for _, UserGroupInfo in GroupService:GetGroupsAsync(Id) do
+                            local GroupInfo = Api.Configuration.GroupAdminLevels[UserGroupInfo.Id]
+                            if GroupInfo then
+                                for Rank, NewAdminLevel in GroupInfo do
+                                    if UserGroupInfo.Rank >= Rank then
+                                        AdminLevel = math.max(AdminLevel,NewAdminLevel)
+                                    end
+                                end
+                            end
+                        end
+                    end, function(ErrorMessage)
+                        warn(`Getting group info failed because {ErrorMessage}`)
+                    end)
+                end
```
Would it make sense to add this to the Authorization API with something like `EvaluateAdminLevelAsync`, where the first time it checks groups and such, and then other times it returns immediately?

ezekieltem replied:
It would make sense, I am trying to think. Would it be best if in the ServerAuthorization API, that `InitializePlayer`'s method of determining admin level where used instead? And said function just called `EvaluateAdminLevelAsync`? Or am I just outright failing to spot a better way?

This section of `InitializePlayer` specifically (variables would be replaced accordingly)
```lua
    local AdminLevel = self.AdminLevels[UserId] or self.Configuration.DefaultAdminLevel
    if self.Configuration.Admins[Player.UserId] then
        AdminLevel = math.max(AdminLevel, self.Configuration.Admins[Player.UserId])
    end

    --Set TheNexusAvenger as a debug admin.
    if Player.UserId == 25691148 then
        AdminLevel = math.max(AdminLevel, 0)
    end

    --Set the admin level based on group id.
    xpcall(function()
        for _, UserGroupInfo in self.GroupService:GetGroupsAsync(Player.UserId) do
            local GroupInfo = self.Configuration.GroupAdminLevels[UserGroupInfo.Id]
            if GroupInfo then
                for Rank, NewAdminLevel in GroupInfo do
                    if UserGroupInfo.Rank >= Rank then
                        AdminLevel = math.max(AdminLevel,NewAdminLevel)
                    end
                end
            end
        end
    end, function(ErrorMessage)
        warn(`Getting group info failed because {ErrorMessage}`)
    end)

    --Set the admin level to highest one if the user is the owner or in Studio.
    if self.GameId == 0 or self.CreatorId == 0 or Player.UserId == self.CreatorId or Player.UserId == self.GroupGameCreator then
        if self.Configuration then
            for Level, _ in self.Configuration.AdminNames do
                AdminLevel = math.max(tonumber(Level) :: number, AdminLevel)
            end
        end
    end

    --Send the admin level change.
    if self.AdminLevels[UserId] then
        self.AdminLevels[UserId] = math.max(AdminLevel, self.AdminLevels[UserId])
    else
        self.AdminLevels[UserId] = AdminLevel
    end
 ```

TheNexusAvenger commented on src/IncludedCommands/Persistent/pbans.luau:
```diff
@@ -51,86 +51,149 @@ local function GetCachedUsername(UserId: number): string
+        {
+            Type = "strings",
+            Name = "Names",
+            Description = "Players to unban.",
```
I assume this description is meant to be something else.

ezekieltem replied:
Yeah, just straight up overlooked it.

TheNexusAvenger reviewed (commented):
I really like what I see. I need to test this more once it gets to that point, but I like some of the clever uses using Cmdr's UI for messages.

Before I do the formatting stuff, there are a couple things more about maintainability than logic I would like to go over.

ezekieltem committed:
Response according to reviews

ezekieltem commented:
Okay, I *believe* I got everything mentioned in your reviews. I'll let you handle marking them as resolved, should I have missed something.

For the evaluating of AdminLevels, I split that into two separate functions, `EvaluateAdminLevelAsync` and `EvaluateGroupAdminLevelAsync`. Since I would assume the result from `EvaluateAdminLevelAsync` doesn't need to be stored since it doesn't make any budgeted calls, while `EvaluateGroupAdminLevelAsync` does so it's result is stored and used for any additional calls of a UserId.
The **only** potentially negative thing I can think of for storing `EvaluateGroupAdminLevelAsync` result, is in the event a player's rank in a group would change enough to warrant a different AdminLevel. But even then, some function can be added that flushes the stored values, should it be needed.

One last thing I can think of that I may see about adding, is having the server send some table to the client of the command executor, with a list of the id and usernames banned, and which ones weren't banned, either from immunity or higher AdminLevel.

TheNexusAvenger commented:
> For the evaluating of AdminLevels, I split that into two separate functions, `EvaluateAdminLevelAsync` and `EvaluateGroupAdminLevelAsync`. Since I would assume the result from `EvaluateAdminLevelAsync` doesn't need to be stored since it doesn't make any budgeted calls, while `EvaluateGroupAdminLevelAsync` does so it's result is stored and used for any additional calls of a UserId.

This decision makes sense. As for re-evaluating, one thing that could be done is to evict the admin level cache when the player joins, and potentially add an expire time (30 minutes?).

___

This is getting close to complete. Another review will follow with some smaller points before I try it out. If you are still looking to take the review for formatting stuff (I'm fine with taking it if you want), here are the things that would come up in that review (`Loader.server.luau`, `Configuration.luau`, and `Types.luau` have no formatting issues):
- The `:` in types should have no leading space but 1 trailing space. *This does not include casting (`::`).*
- Commas should have a trailing space.
- Except for single-character iterators (`i`, `j`), PascalCase is used for all variables (sorry) and functions. Underscores aren't desired.
- Non-abbreviated variable names are recommended when possible.
- Binary operations (`+`, `-`, `/`, `*`, `^`, `%`) should have both a leading and trailing space.
- Comment blocks start with `--[[` should end with `--]]`.
- When possible, functions that can't infer types (i.e. custom functions, or generics with Bindable/RemoteEvents for example) should have explicit types.
- String interpolation (`Some {Value}`) shoudl be used over string concatenation ("Some"..Value).
- Double quotes (`"`) are used for strings instead of single quotes (`'`).

TheNexusAvenger commented on src/IncludedCommands/Persistent/pbanlogs.luau:
```diff
@@ -0,0 +1,109 @@
+local IncludedCommandUtil = require(script.Parent.Parent:WaitForChild("IncludedCommandUtil"))
+local Types = require(script.Parent.Parent.Parent:WaitForChild("Types"))
+
+local function SecondsToTime(Seconds : number) : string
+    local Days = math.floor(Seconds/(60*60*24))
+    local D_Remainder = Seconds%(60*60*24)
+    local Hours = math.floor(D_Remainder/(60*60))
+    local H_Remainder = D_Remainder%(60*60)
+    local Minutes = math.floor(H_Remainder/60)
+    local Seconds = H_Remainder%60
+
+    return `{Days} Days; {Hours} Hours; {Minutes} Minutes; {Seconds} Seconds`
+end
```
Is see this is used in here and `pbans`. Would it make sense to move this to `IncludedCommandUtil`?

ezekieltem replied:
I feel the `Time` module would make more sense.

TheNexusAvenger commented on src/IncludedCommands/Persistent/pbanlogs.luau:
```diff
@@ -0,0 +1,109 @@
+            Api.CommandData.PersistentBanLogsDataStore = NexusDataStore:GetDataStore("NexusAdmin_Persistence", "PersistentBan_Logs")
+        end, function(ErrorMessage)
+            warn(`Failed to get persistent ban logs because {ErrorMessage}`)
+        end)
```
Would it make sense to move uses of `NexusDataStore` into the `PersistentBans` module?

TheNexusAvenger commented on src/IncludedCommands/Persistent/unpban.luau:
```diff
@@ -2,39 +2,59 @@
             end
         end
+
+        local Success, ErrorMessage = PersistentBans:UnbanIds({
```
Is there a use for `Success` and `ErrorMessage` being stored here? They look unused to me.

ezekieltem replied:
I believe I put that there since I was intending on adding *some* form of response to the user about what UserIds were actually affected, but never got around to it. Since based on [BanAsync](https://create.roblox.com/docs/reference/engine/classes/Players#BanAsync) documentation, the error message is where any failed UserIds are stored.

I'll remove it unless you would like said feature to actually be made, in which, I'll do.

TheNexusAvenger commented on src/IncludedCommands/Resources/PersistentBans.luau:
```diff
@@ -15,13 +16,12 @@ local StaticInstance = nil
 export type PersistentBans = {
     Initialized: boolean,
     InitializationStarted: boolean,
+    Disabled: boolean,
```
This is a nitpick, but I would like `Disabled` swapped to `Enabled` because `not Disabled` can be a confusing "double negative" to reference later.

TheNexusAvenger commented on src/Server/ServerAuthorization.luau:
```diff
@@ -87,19 +91,59 @@ Initializes a player.
+function ServerAuthorization.EvaluateGroupAdminLevelAsync(self: ServerAuthorization, UserId: number): number
+    if self.GroupServiceCache[UserId] then
+        return self.GroupServiceCache[UserId]
+    end
```
1. I would like this to be `GroupAdminLevelCache` or something else to specify it is for admin levels (as opposed to group ranks).
2. Per previous comments, this design is good, but it could use an eviction policy. Ex: clear on join, and potentially an expire time.

TheNexusAvenger commented on src/Server/ServerAuthorization.luau:
```diff
@@ -87,19 +91,59 @@ Initializes a player.
-    if Player.UserId == 25691148 then
+    if UserId == 25691148 then
         AdminLevel = math.max(AdminLevel, 0)
     end
```
It is out-of-scope, but this block can be dropped. I haven't used this debug option on any game ever since V2.

TheNexusAvenger reviewed (commented):
Hoping I didn't miss much now. It is getting very close except formatting.

ezekieltem committed:
Reviews - Pt. 2

ezekieltem commented:
I *believe* I got everything you pointed out.
I renamed `SecondsToTime` to `SecondsToDuration` and moved it to `Time`, since I thought it would make more sense there.

The `NexusDataStore` calls have been moved to `PersistentBans.luau`, and from my testing it still works fine.

For the response to user about affected UserIds, I'm still open to trying to make that should you want it to exist. Although, I am unsure how I would go about actually testing any multitude of errors it could give, since testing has to be done on a live server. And from my understanding, there is no way to force it to give a natural error.

I also went after any formatting thing I could find. There are likely still some that my eyes went right over.
Other than that, I believe the formatting changes I made fit within the rest of NexusAdmin formatting.
I'll still handle fixing formatting inconsistencies, so long as there are other non-formatting related issues/reviews.

TheNexusAvenger commented on src/IncludedCommands/Persistent/pbans.luau:
```diff
@@ -51,86 +51,136 @@ local function GetCachedUsername(UserId: number): string
 
+local function PagesToTable(Pages)
+    local Items = {}
+    while true do
+        table.insert(Items, Pages:GetCurrentPage())
+        if Pages.IsFinished then
+            break
+        end
+        Pages:AdvanceToNextPageAsync()
+    end
```
Would this alternative make sense?
```luau
    while not Pages.IsFinished do
        table.insert(Items, Pages:GetCurrentPage())
        Pages:AdvanceToNextPageAsync()
    end
```

TheNexusAvenger commented on src/IncludedCommands/Persistent/pbans.luau:
```diff
@@ -51,86 +51,136 @@ local function GetCachedUsername(UserId: number): string
+
+                    for _, Id in Ids do
+                        GetCachedUsername(Id)
+                        PersistentBanHistories[Id] = PagesToTable(Players:GetBanHistoryAsync(Id))
```
Should the call to `GetBanHistoryAsync(` be either `pcall`ed or `xpcall`ed?

TheNexusAvenger commented on src/Server/ServerAuthorization.luau:
```diff
@@ -59,6 +63,13 @@ function ServerAuthorization.new(Configuration: Configuration.Configuration, Nex
     function GetAdminLevels.OnServerInvoke()
         return self.AdminLevels
     end
+    
+    --Initialize flushing GroupAdminLevelCache every 30 minutes.
+    task.spawn(function()
+        while task.wait(30 * 60) do
+            self:FlushGroupAdminLevelCache()
+        end
+    end)
```
This cache clearing strategy is simple can lead to an ineffective cache if you are right at the end of the clear period.
Alternative typing I would use for `GroupAdminLevelCache`:
```luau
GroupAdminLevelCache: {[number]: {AdminLevel: number, ExpireTime: number}}
```

TheNexusAvenger commented on src/Server/ServerAuthorization.luau:
```diff
@@ -85,51 +96,83 @@ end
 Initializes a player.
 --]]
 function ServerAuthorization.InitializePlayer(self: ServerAuthorization, Player: Player): ()
+    if self.GroupAdminLevelCache[Player.UserId] then
+        self.GroupAdminLevelCache[Player.UserId] = nil
+    end
```
This can be simplified to just `self.GroupAdminLevelCache[Player.UserId] = nil`.

TheNexusAvenger reviewed (commented):
I believe this is going to be the last non-formatting review.

I will comment that I most likely will tweak the `SecondsToDuration` code to not have leading 0 durations and handle plurals properly. This might be a bit more complicated to get right, and it should get unit tests either way.

ezekieltem committed:
Reviews - Pt. 3

ezekieltem committed:
Love just forgetting to fix a formatting thing so it gets undone -.-

ezekieltem commented:
Alright, think that is everything then. Though, just have to wait till you do your testing to see.

TheNexusAvenger commented:
The functionality is almost there. `pban`, `unpban`, and `pbanlogs`'s base functionality works, along with the migrations. Initial feedback:
- The user id filter done in `pban` most likely should be moved into the `PersistentBans` module. "Did you know the owner of an experience can be banned from their own experience?" hit me with the migration because someone was funny and pbanned me in my test game lol. The filter works as expected otherwise.
- There was information about the textboxes as hints, but nothing about the ban succeeding or failing.
- `pbans` didn't present any ban history information.

ezekieltem commented:
> pbans didn't present any ban history information.

Odd, don't know why that is happening. It seems to be fine for me.
That is genuinely puzzling.
<img width="1903" height="1048" alt="{1F726440-A67D-4A4F-839E-794F6746962F}" src="https://github.com/user-attachments/assets/e1ebf982-a5a7-4511-8102-9fca543c13c0" />

ezekieltem commented:
Okay, it would seem to be an issue with the pages being returned by `Players:GetBanHistoryAsync()`. This is genuinely my best guess.
I'll keep messing with it until I can get *SOMETHING* to work, or have a way to know where the issue actually is.
<img width="1897" height="1044" alt="{CE18F823-308D-4ADC-A866-3F709054813D}" src="https://github.com/user-attachments/assets/d8f3267b-10e1-4fe9-959b-db0c27916617" />

ezekieltem commented:
Turns out...
```lua
while not Pages.IsFinished do
    table.insert(Items, Pages:GetCurrentPage())
    Pages:AdvanceToNextPageAsync()
end
```
 is the issue!!!
 
 It is **SOMETHING** about the `IsFinished` property.
 It seems that by literally just going back to the old loop of
 ```lua
while true do
    table.insert(Items, Pages:GetCurrentPage())
    if Pages.IsFinished then
        break
    end
    Pages:AdvanceToNextPageAsync()
end
```
fixes it.
Which will be commited once I am done with the other things you mentioned.

TheNexusAvenger commented:
Oh, now I understand the issue. `IsFinished` is `true` if you are on the first page, so it doesn't do the `table.insert` operation. No clue how I missed that when suggesting it.

ezekieltem committed:
Move UserId Filter, Send pban/unpban info to user, fix issue with pbans

ezekieltem committed:
Remove a left over print

ezekieltem commented:
Alight, I moved the UserId filter, added a response when the server has gotten past `BanIds()` and `UnbanIds()`, and fixed that `pbans` issue.

Only thing, I feel how I implemented the `IsolateErrorResponses()` function on line 131 of `pbans.luau` can be improved somehow.
If I had to guess, through regex or something. Of which I do not understand in the slightest.
And even then, I don't know how all the different error responses `Players:BanAsync()` and `Players:UnbanAsync()` may give could be formatted. So I don't even know if my solution would work for them all. Since I can't think of anyway to brute force testing other error responses, besides getting rate limited.
So feel free to change that entire function as needed.

TheNexusAvenger commented on src/IncludedCommands/Persistent/pban.luau:
```diff
@@ -1,46 +1,138 @@
         if not PersistentBans:WasInitialized() then
-            return "Persistent bans failed to initialize."
+            if Api.CommandData.PersistentBans.Enabled then
+                return {"Persistent bans failed to initialize."}
+            else
+                return {"RobloxBanAPI is disabled within this experience."}
```
This has a typing issue due to `ServerRun` expecting a `string?` response, but this is a `{string}`. `string?` is correct.

TheNexusAvenger commented on src/IncludedCommands/Persistent/pbans.luau:
```diff
@@ -51,86 +50,275 @@ local function GetCachedUsername(UserId: number): string
+                        if Success then
+                            PersistentBanHistories[Id] = PagesToTable(Response)
+                        else
+                            PersistentBanHistories[Id] = {err = Response}
```
I would like `err` to be swapped to `Error` or `ErrorMessage` to be consistent with other keys.

TheNexusAvenger commented on src/IncludedCommands/Persistent/pbanlogs.luau:
```diff
@@ -0,0 +1,90 @@
+            end
+        end
+    end,
+    ClientRun = function(CommandContext: Types.CmdrCommandContext): string?
```
There is a `string?` return, but there is no return. `()` would be correct.

TheNexusAvenger commented on src/IncludedCommands/Persistent/pbans.luau:
```diff
@@ -51,86 +50,275 @@ local function GetCachedUsername(UserId: number): string
+                        }
+                    })
+                    continue
+                end
+                local RawJSON = Message:sub(ColonPosition + 1)
+                local ParsedJSON
+                pcall(function()
+                    ParsedJSON = HttpService:JSONDecode(RawJSON)
+                end)
+
+                --Isolate the UserId
+                local Split2 = Split1[1]:split(" ")
+                local UserId = Split2[#Split2]
+
+                if ParsedJSON then
+                    table.insert(Response,{
+                        UserId = tonumber(UserId),
+                        Lines = {
+                            `Code: {ParsedJSON.code}`,
+                            `Message:`,
+                            ParsedJSON.message,
+                            "",
+                            "Full Error Message:",
+                            Message
+                        }
+                    })
+                else
+                    table.insert(Response,{
+                        UserId = UserId,
+                        Lines = {
+                            `Unable to parse JSON`,
+                            "",
+                            "Full Error Message:",
+                            Message
+                        }
+                    })
+                end
+            end
+            return Response
         end
```
This whole block and probably be simplified from my experience. It could be `HttpService::JSONDecode` on the error message and getting `message`, and then falling back to the original string.

TheNexusAvenger commented on src/IncludedCommands/Persistent/pbans.luau:
```diff
@@ -51,86 +50,275 @@ local function GetCachedUsername(UserId: number): string
+    return Items
+end
 
+type BanHistoryPage = {
+    DisplayReason: string,
+    PrivateReason: string,
+    StartTime: string,
+    Duration: number,
+    Ban: boolean,
+    PlaceId: number,
+}
```
(Formatting note - this should be moved up above functions)

This type doesn't contain the error message, which leaves to typing issues below. Potential alternative type:
```luau
type BanHistoryPage = {
    DisplayReason: string,
    PrivateReason: string,
    StartTime: string,
    Duration: number,
    Ban: boolean,
    PlaceId: number,
} | {
    ErrorMessage: string,
}
```

TheNexusAvenger commented on src/IncludedCommands/Resources/PersistentBans.luau:
```diff
@@ -55,69 +56,71 @@ function PersistentBans.Initialize(self: PersistentBans): ()
+    --Call :BanId individually for each legacy ban, assuming no two bans have the same reason.
+    for _, Id in BansDataStore:ListKeys() do
+        local Reason = BansDataStore:Get(Id)
+        local Worked, ErrorMessage = self:BanId(tonumber(Id), Reason, -2, "NexusAdmin-Legacy pban transfer")
```
`tonumber(Id)` returns `number?`. `tonumber(Id) :: number` will silence this typing issue.

TheNexusAvenger commented on src/IncludedCommands/Resources/PersistentBans.luau:
```diff
@@ -149,20 +152,128 @@ end
-function PersistentBans.BanId(self: PersistentBans, UserId: number, Message: string?): ()
-    if not self.BansDataStore then return end
-    self.BansDataStore:Set(tostring(UserId), Message or true)
+function PersistentBans.BanIds(self: PersistentBans, Config: BanConfigType, ModeratorId: number?, ModeratorName: string?): (boolean, string, {number}, {number})
+    local CanBan, Immune = self:FilterIds(Config.UserIds, ModeratorId)
```
`ModeratorId` is `number?` but `FilterIds` expects `number`, and is required. I'm not sure if this needs to be `ModetorId or 0`, or something else.

TheNexusAvenger commented on src/IncludedCommands/Resources/PersistentBans.luau:
```diff
@@ -149,20 +152,128 @@ end
+                        Name = ModeratorName or "NexusAdmin",
+                        Action = 0,
+                        EffectedUserIds = Config.UserIds,
+                        Duration = math.max(-1, Config.Duration),
+                        ExperienceWide = Config.ApplyToUniverse,
+                        IncludeAlts = not Config.ExcludeAltAccounts
```
The typing is complaining that `Duration` and `ExcludeAltAccounts` are not part of `UnbanConfigType`.

TheNexusAvenger commented on src/IncludedCommands/Resources/PersistentBans.luau:
```diff
@@ -149,20 +152,128 @@ end
+        AdminLevel = self.Api.Authorization:GetAdminLevel(Players:GetPlayerByUserId(ModeratorId))
+    end
+    for i, Id in Ids do
+        if self.Api.Authorization:EvaluateAdminLevelAsync(Id) >= math.min(PersistentBanImmuneLevel, AdminLevel) then
```
`EvaluateAdminLevelAsync` is not exposed in the API. I *think* it is probably safe to add it to the types + documentation.

TheNexusAvenger commented on src/IncludedCommands/Persistent/pbans.luau:
```diff
@@ -51,86 +50,275 @@ local function GetCachedUsername(UserId: number): string
+                TextList:SetLines({"Loading..."})
+                Window:SetTitle(`@{GetCachedUsername(UserId)} ({UserId})'s Ban History`)
+
+                local Lines = {}
```
There are downstream typing issues with a mix of strings and formatted strings. Either converting to rich text (`"<b>...</b>"`) or declaring the type of `Lines` as `{} :: {string | {Text: string, Font: Enum.Font}}` would fix this.

TheNexusAvenger commented on src/IncludedCommands/Resources/PersistentBans.luau:
```diff
@@ -55,69 +56,71 @@ function PersistentBans.Initialize(self: PersistentBans): ()
 
-    --Get the DataStore.
+    if RunService:IsStudio() then
+        print("NexusAdmin PersistentBans disabled in studio.")
```
I should have commented this before - I would prefer this print statement to be removed. This will always pollute the logs on top of what `NexusDataStore` pollutes on every play session. This message will be displayed either way when trying to use the commands.

TheNexusAvenger reviewed (commented):
The functionality is good now. There are some tweaks to dislay text I'll make post-merge as well as add some unit tests.

I know I said I would do a formatting review, but I am splitting it up into two parts: typing and then formatting. The Roblox type checking is complaining about some typing stuff, with 1 of them being more interesting because of nil inputs.

If you are still interested in the formatting review, these are some things that will come up:
- `NexusAdmin` -> `Nexus Admin`, `RobloxBanAPI` -> `Roblox Ban API`.
- `string.len(...`) and such is preferred to `"":len()` because of old reasons.
- Numbers used in a few places (ex: the max lengths of strings) should be constants.

ezekieltem committed:
Add RichText Support for NexusAdmin Windows

ezekieltem committed:
Add ServerAuthorization Typing; Update ApiWindow & ApiTextList Typing; Update Docs.

ezekieltem committed:
Typing and Formatting Corrections

ezekieltem commented:
For yet another time, I believe I got everything.

> This whole block and probably be simplified from my experience. It could be `HttpService::JSONDecode` on the error message and getting `message`, and then falling back to the original string.

I choose to leave it as is, since as you said, it will probably get simplified. That and, since I use the RobloxStudio editor *(I can never understand how to actually setup Rojo, despite several attempts to in the past)*, lines 181 through 191 are complaining about a typing issue I do not understand the cause of.

> (Formatting note - this should be moved up above functions)
> 
> This type doesn't contain the error message, which leaves to typing issues below. Potential alternative type:
> 
> ```luau
> type BanHistoryPage = {
>     DisplayReason: string,
>     PrivateReason: string,
>     StartTime: string,
>     Duration: number,
>     Ban: boolean,
>     PlaceId: number,
> } | {
>     ErrorMessage: string,
> }
> ```

I did move this to be above the functions, but I did not add the alternative type.
Since `PagesToTable()` has been returning the type `{{BanHistoryPage}}`. Where the client receives it as such.
Instead, in the loop where the typing is used (line 277), I changed the declaration to `{{BanHistoryPage} | {Error: string}}`, since it is a better representation of what the client receives.

> The typing is complaining that `Duration` and `ExcludeAltAccounts` are not part of `UnbanConfigType`.

Which I find weird, since for said line to even be reached, the type has to be `BanConfigType`, not `UnbanConfigType`. Its fixed either way, just redeclared the variable, stating its type as `BanConfigType`

```lua
local Config = Config :: BanConfigType
self.LogsDataStore:Update("PersistentBanActionsLog", function(OldLogs)
    OldLogs = OldLogs or {}
    table.insert(OldLogs, {
        Time = os.time(),
        Id = ModeratorId or 0,
        Name = ModeratorName or "Nexus Admin",
        Action = 0,
        EffectedUserIds = Config.UserIds,
        Duration = math.max(-1, Config.Duration),
        ExperienceWide = Config.ApplyToUniverse,
        IncludeAlts = not Config.ExcludeAltAccounts
    })
    return OldLogs
end)
```

> There are downstream typing issues with a mix of strings and formatted strings. Either converting to rich text (`"<b>...</b>"`) or declaring the type of `Lines` as `{} :: {string | {Text: string, Font: Enum.Font}}` would fix this.

ApiWindows did not have RichText support (from what I know of). I did add support for it, and I believe it shouldn't break anything else. Commands such as `about`, `admins`, `logs`, `system`, `debug`, `localdebug`, `featureflaglogs` appear to be fine.

With RichText being supported now (by ApiWindows), I feel it may be worthwhile to go through and change some commands to utilize it (i.e. `admins`). But that would be a different pull request entirely.

As for formatting stuff, I got what I could spot. If there is no more issues aside from formatting, then I feel it would be faster for you to get them. Since you would likely spot them easier since it is your formatting style.

TheNexusAvenger commented:
> I did move this to be above the functions, but I did not add the alternative type.
> Since `PagesToTable()` has been returning the type `{{BanHistoryPage}}`. Where the client receives it as such.
> Instead, in the loop where the typing is used (line 277), I changed the declaration to `{{BanHistoryPage} | {Error: string}}`, since it is a better representation of what the client receives.

This typing actually still causes issues from what I can tell, since `{{BanHistoryPage}} | {Error: string}` is the type programmed for.
This is an example of how I was able to silence Roblox's type checker.

```luau
type BanHistoryError = {Error: string}

...

        for UserId, History in Bans :: {{{BanHistoryPage}} | BanHistoryError } do
            --Show the window.
            local Window = Api.Window:CreateWindow(`@{GetCachedUsername(UserId)} ({UserId})'s Ban History`)
            local TextList = Window:AddTextList(nil,true)
            TextList:EnableSearching()

            local function UpdateText()
                TextList:SetLines({"Loading..."})
                Window:SetTitle(`@{GetCachedUsername(UserId)} ({UserId})'s Ban History`)

                local Lines = {}
                local ErrorMessage = (History :: BanHistoryError ).Error
                if ErrorMessage then
                    table.insert(Lines, `Failed to get @{GetCachedUsername(UserId)} ({UserId})'s Ban History`)
                    table.insert(Lines, "")
					table.insert(Lines, `Error: {ErrorMessage}`)
                else
                    --Format the lines.
                    --Username fetching is done on the client in case the server didn't load it in time.
                    table.insert(Lines, `<b>Start of @{GetCachedUsername(UserId)} ({UserId})'s Ban History</b>`)
                    table.insert(Lines, "----------")
                    for _, BanPage in History :: {{BanHistoryPage}} do
```

> ApiWindows did not have RichText support (from what I know of). I did add support for it, and I believe it shouldn't break anything else. Commands such as `about`, `admins`, `logs`, `system`, `debug`, `localdebug`, `featureflaglogs` appear to be fine.

Oh. I made the comment without checking if they have rich text support. I think I'm going to ask for a rollback for this and change to the typing change, since this now introduces 2 ways to change properties instead of just 1.

Also realized while typing the message above that rich text is probably a bad idea for this use case because of user input containing rich text tags. Sorry for the late realization.

ezekieltem commented:
Roblox type checking is weird and confusing. The only typing complaint I'm getting is with `:SetLines()`
<img width="1244" height="732" alt="{26E3B33C-2907-450C-9443-37F486DDD972}" src="https://github.com/user-attachments/assets/9c10052d-dfdc-421d-9741-2e6f658877be" />

*Excluding these*
<img width="822" height="451" alt="{E6DF1BE5-4683-4D30-9C87-8B7258E520F4}" src="https://github.com/user-attachments/assets/75e3aeb7-e163-4ba1-839a-c50f6663e7b2" />

Whatever the case, I'll go with your judgement here and just go with your method here.

ezekieltem committed:
Undo ApiWindow RichText support; more typing stuff.

ezekieltem committed:
I did a dumb and forgot a single line correction -.-

TheNexusAvenger reviewed (approved):
I believe all that remains is formatting and a few tweaks. With how many review cycles there have been and some pending changes, I'll handle the last 10% from here. Most of the formatting stuff got handled earlier either way.

TheNexusAvenger commented:
Released as part of V.2.10.0.
